Sprint 05

pavelbelenkow opened the pull request:
Requirements for the 5th sprint for the quiz app have been implemented.

pavelbelenkow committed:
Created models for:
QuizQuestion
QuizResultsViewModel
QuizStepViewModel

pavelbelenkow committed:
Created:
Single class QuestionFactory for generating questions
Implementation of the interface via QuestionFactoryProtocol
QuestionFactory delegate

pavelbelenkow committed:
Refactoring MovieQuizViewController:
controller class implements the QuestionFactoryDelegate
private structures have been removed and moved to different models
added weak link for alert
updated methods with new intro
didReceiveNextQuestion delegate protocol method is implemented to work with the requested question

Added properties:
questionAmount - total number of quiz questions;
questionFactory - for contacting the controller for questions for the quiz;
currentQuestion - the current question that the user sees

pavelbelenkow committed:
Alert is placed in a single AlertModel
Added AlertPresenter class to represent the alert
Added AlertPresenterProtocol to implement the AlertPresenter interface

pavelbelenkow committed:
Implemented and placed in single files:
GameRecord type to save the current record in the quiz
StatisticServiceImplementation class for maintaining and saving statistics parameters in UserDefaults:
number of completed games;
best game;
average accuracy of correct answers.
The StatisticService protocol, which sets parameters for the implementation of:
number of completed games;
best game;
average accuracy of correct answers;
method for saving the results of the current game.

pavelbelenkow committed:
Added to the controller class:
statisticService variable of the StatisticService protocol type
Initializer of the StatisticServiceImplementation class in the viewDidLoad method
Updated message for alert that shows statistics from StatisticServiceImplementation

pavelbelenkow committed:
Changed the time format to 24 hour

mmbabaev commented on MovieQuiz/Models/AlertModel.swift:
```diff
@@ -0,0 +1,8 @@
+import UIKit
+
+struct AlertModel {
+    let title: String?
```
зачем делать все поля опциональными?

pavelbelenkow replied:
Оставил опциональным только completion

mmbabaev commented on MovieQuiz/Models/GameRecord.swift:
```diff
@@ -0,0 +1,19 @@
+    let total: Int
+    let date: Date
+    
+    static func < (lhs: GameRecord, rhs: GameRecord) -> Bool {
```
достаточно только этой функции, остальные автоматом вытекают

pavelbelenkow replied:
Удалены ненужные методы сравнения

mmbabaev commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,28 @@
+class AlertPresenter: AlertPresenterProtocol {
+    private weak var viewController: UIViewController?
+    
+    init(viewController: UIViewController?) {
```
в принципе нет смысла передавать опционал

pavelbelenkow replied:
Тип параметра изменен на неопциональный

mmbabaev commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,88 @@
+        }
+        
+        set {
+            guard let data = try? JSONEncoder().encode(newValue) else {
```
можно вынести JSONEncoder() / JSONDecoder в отдельное поле

pavelbelenkow replied:
Создал два приватных экземпляра энкодера и декодера

pavelbelenkow committed:
AlertModel some properties changed to non-optional
Removed unnecessary GameRecord comparison methods
Initializer parameter type has been changed to non-optional in AlertPresenter
Added private properties for decoder and encoder in StatisticServiceImplementation

pavelbelenkow commented:
Можно мерджить?
